File: mobile/mobile_central/test/services/modules/warehouses/app/use_cases_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/warehouses/app/use_cases.dart';
import 'package:mobile_central/services/modules/warehouses/domain/entities.dart';
import 'package:mobile_central/services/modules/warehouses/domain/ports.dart';
import 'package:mobile_central/shared/types/paginated_response.dart';

// --- Manual Mock ---

class MockWarehouseRepository implements IWarehouseRepository {
  final List<String> calls = [];

  PaginatedResponse<Warehouse>? getWarehousesResult;
  WarehouseDetail? getWarehouseByIdResult;
  Warehouse? createWarehouseResult;
  Warehouse? updateWarehouseResult;
  List<WarehouseLocation>? getLocationsResult;
  WarehouseLocation? createLocationResult;
  WarehouseLocation? updateLocationResult;

  Exception? errorToThrow;

  GetWarehousesParams? capturedParams;
  int? capturedId;
  int? capturedBusinessId;
  CreateWarehouseDTO? capturedCreateData;
  int? capturedUpdateId;
  UpdateWarehouseDTO? capturedUpdateData;
  int? capturedDeleteId;
  int? capturedWarehouseId;
  CreateLocationDTO? capturedCreateLocationData;
  int? capturedLocationId;
  UpdateLocationDTO? capturedUpdateLocationData;
  int? capturedDeleteWarehouseId;
  int? capturedDeleteLocationId;

  @override
  Future<PaginatedResponse<Warehouse>> getWarehouses(GetWarehousesParams? params) async {
    calls.add('getWarehouses');
    capturedParams = params;
    if (errorToThrow != null) throw errorToThrow!;
    return getWarehousesResult!;
  }

  @override
  Future<WarehouseDetail> getWarehouseById(int id, {int? businessId}) async {
    calls.add('getWarehouseById');
    capturedId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getWarehouseByIdResult!;
  }

  @override
  Future<Warehouse> createWarehouse(CreateWarehouseDTO data, {int? businessId}) async {
    calls.add('createWarehouse');
    capturedCreateData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return createWarehouseResult!;
  }

  @override
  Future<Warehouse> updateWarehouse(int id, UpdateWarehouseDTO data, {int? businessId}) async {
    calls.add('updateWarehouse');
    capturedUpdateId = id;
    capturedUpdateData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return updateWarehouseResult!;
  }

  @override
  Future<void> deleteWarehouse(int id, {int? businessId}) async {
    calls.add('deleteWarehouse');
    capturedDeleteId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
  }

  @override
  Future<List<WarehouseLocation>> getLocations(int warehouseId, {int? businessId}) async {
    calls.add('getLocations');
    capturedWarehouseId = warehouseId;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getLocationsResult!;
  }

  @override
  Future<WarehouseLocation> createLocation(int warehouseId, CreateLocationDTO data, {int? businessId}) async {
    calls.add('createLocation');
    capturedWarehouseId = warehouseId;
    capturedCreateLocationData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return createLocationResult!;
  }

  @override
  Future<WarehouseLocation> updateLocation(int warehouseId, int locationId, UpdateLocationDTO data, {int? businessId}) async {
    calls.add('updateLocation');
    capturedWarehouseId = warehouseId;
    capturedLocationId = locationId;
    capturedUpdateLocationData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return updateLocationResult!;
  }

  @override
  Future<void> deleteLocation(int warehouseId, int locationId, {int? businessId}) async {
    calls.add('deleteLocation');
    capturedDeleteWarehouseId = warehouseId;
    capturedDeleteLocationId = locationId;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
  }
}

// --- Helpers ---

Warehouse _makeWarehouse({int id = 1, String name = 'TestWarehouse'}) {
  return Warehouse(
    id: id,
    businessId: 1,
    name: name,
    code: 'WH-001',
    address: 'Addr',
    city: 'City',
    state: 'State',
    country: 'CO',
    zipCode: '110111',
    phone: '123',
    contactName: 'CN',
    contactEmail: 'ce@t.com',
    isActive: true,
    isDefault: false,
    isFulfillment: false,
    company: 'Co',
    firstName: 'F',
    lastName: 'L',
    email: 'e@t.com',
    suburb: 'S',
    cityDaneCode: '11001',
    postalCode: '110111',
    street: 'St',
    createdAt: '2026-01-01',
    updatedAt: '2026-01-01',
  );
}

WarehouseDetail _makeWarehouseDetail({int id = 1}) {
  return WarehouseDetail(
    id: id,
    businessId: 1,
    name: 'Detail WH',
    code: 'DWH',
    address: 'A',
    city: 'C',
    state: 'S',
    country: 'CO',
    zipCode: '110111',
    phone: '1',
    contactName: 'CN',
    contactEmail: 'ce@t.com',
    isActive: true,
    isDefault: false,
    isFulfillment: false,
    company: 'Co',
    firstName: 'F',
    lastName: 'L',
    email: 'e@t.com',
    suburb: 'S',
    cityDaneCode: '11001',
    postalCode: '110111',
    street: 'St',
    createdAt: '2026-01-01',
    updatedAt: '2026-01-01',
    locations: [],
  );
}

WarehouseLocation _makeLocation({int id = 1, String name = 'Shelf A'}) {
  return WarehouseLocation(
    id: id,
    warehouseId: 1,
    name: name,
    code: 'SA',
    type: 'shelf',
    isActive: true,
    isFulfillment: false,
    createdAt: '2026-01-01',
    updatedAt: '2026-01-01',
  );
}

Pagination _makePagination() {
  return Pagination(
    currentPage: 1,
    perPage: 10,
    total: 1,
    lastPage: 1,
    hasNext: false,
    hasPrev: false,
  );
}

// --- Tests ---

void main() {
  late MockWarehouseRepository mockRepo;
  late WarehouseUseCases useCases;

  setUp(() {
    mockRepo = MockWarehouseRepository();
    useCases = WarehouseUseCases(mockRepo);
  });

  group('getWarehouses', () {
    test('delegates to repository and returns result', () async {
      final expected = PaginatedResponse<Warehouse>(
        data: [_makeWarehouse()],
        pagination: _makePagination(),
      );
      mockRepo.getWarehousesResult = expected;
      final params = GetWarehousesParams(page: 1, pageSize: 10);

      final result = await useCases.getWarehouses(params);

      expect(result.data.length, 1);
      expect(result.data[0].name, 'TestWarehouse');
      expect(mockRepo.calls, ['getWarehouses']);
      expect(mockRepo.capturedParams, params);
    });

    test('passes null params to repository', () async {
      mockRepo.getWarehousesResult = PaginatedResponse<Warehouse>(
        data: [],
        pagination: _makePagination(),
      );

      await useCases.getWarehouses(null);

      expect(mockRepo.capturedParams, isNull);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Network error');

      expect(() => useCases.getWarehouses(null), throwsException);
    });
  });

  group('getWarehouseById', () {
    test('delegates to repository with correct id', () async {
      mockRepo.getWarehouseByIdResult = _makeWarehouseDetail(id: 42);

      final result = await useCases.getWarehouseById(42);

      expect(result.id, 42);
      expect(mockRepo.capturedId, 42);
      expect(mockRepo.calls, ['getWarehouseById']);
    });

    test('passes businessId to repository', () async {
      mockRepo.getWarehouseByIdResult = _makeWarehouseDetail();

      await useCases.getWarehouseById(1, businessId: 5);

      expect(mockRepo.capturedBusinessId, 5);
    });
  });

  group('createWarehouse', () {
    test('delegates to repository with correct data', () async {
      final dto = CreateWarehouseDTO(name: 'New WH', code: 'NWH');
      mockRepo.createWarehouseResult = _makeWarehouse(id: 10, name: 'New WH');

      final result = await useCases.createWarehouse(dto);

      expect(result.id, 10);
      expect(result.name, 'New WH');
      expect(mockRepo.capturedCreateData, dto);
      expect(mockRepo.calls, ['createWarehouse']);
    });

    test('passes businessId to repository', () async {
      final dto = CreateWarehouseDTO(name: 'N', code: 'C');
      mockRepo.createWarehouseResult = _makeWarehouse();

      await useCases.createWarehouse(dto, businessId: 7);

      expect(mockRepo.capturedBusinessId, 7);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Creation failed');
      final dto = CreateWarehouseDTO(name: 'x', code: 'y');

      expect(() => useCases.createWarehouse(dto), throwsException);
    });
  });

  group('updateWarehouse', () {
    test('delegates to repository with correct id and data', () async {
      final dto = UpdateWarehouseDTO(name: 'Updated', code: 'UPD');
      mockRepo.updateWarehouseResult = _makeWarehouse(id: 5, name: 'Updated');

      final result = await useCases.updateWarehouse(5, dto);

      expect(result.id, 5);
      expect(result.name, 'Updated');
      expect(mockRepo.capturedUpdateId, 5);
      expect(mockRepo.capturedUpdateData, dto);
      expect(mockRepo.calls, ['updateWarehouse']);
    });
  });

  group('deleteWarehouse', () {
    test('delegates to repository with correct id', () async {
      await useCases.deleteWarehouse(7);

      expect(mockRepo.capturedDeleteId, 7);
      expect(mockRepo.calls, ['deleteWarehouse']);
    });

    test('passes businessId to repository', () async {
      await useCases.deleteWarehouse(1, businessId: 2);

      expect(mockRepo.capturedBusinessId, 2);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Delete failed');

      expect(() => useCases.deleteWarehouse(7), throwsException);
    });
  });

  group('getLocations', () {
    test('delegates to repository with correct warehouseId', () async {
      mockRepo.getLocationsResult = [_makeLocation(), _makeLocation(id: 2)];

      final result = await useCases.getLocations(1);

      expect(result.length, 2);
      expect(mockRepo.capturedWarehouseId, 1);
      expect(mockRepo.calls, ['getLocations']);
    });

    test('passes businessId to repository', () async {
      mockRepo.getLocationsResult = [];

      await useCases.getLocations(1, businessId: 3);

      expect(mockRepo.capturedBusinessId, 3);
    });
  });

  group('createLocation', () {
    test('delegates to repository with correct data', () async {
      final dto = CreateLocationDTO(name: 'New Loc', code: 'NL');
      mockRepo.createLocationResult = _makeLocation(id: 20, name: 'New Loc');

      final result = await useCases.createLocation(1, dto);

      expect(result.id, 20);
      expect(result.name, 'New Loc');
      expect(mockRepo.capturedWarehouseId, 1);
      expect(mockRepo.capturedCreateLocationData, dto);
      expect(mockRepo.calls, ['createLocation']);
    });
  });

  group('updateLocation', () {
    test('delegates to repository with correct ids and data', () async {
      final dto = UpdateLocationDTO(name: 'Updated', code: 'UL');
      mockRepo.updateLocationResult = _makeLocation(id: 5, name: 'Updated');

      final result = await useCases.updateLocation(1, 5, dto);

      expect(result.id, 5);
      expect(mockRepo.capturedWarehouseId, 1);
      expect(mockRepo.capturedLocationId, 5);
      expect(mockRepo.capturedUpdateLocationData, dto);
      expect(mockRepo.calls, ['updateLocation']);
    });
  });

  group('deleteLocation', () {
    test('delegates to repository with correct ids', () async {
      await useCases.deleteLocation(1, 5);

      expect(mockRepo.capturedDeleteWarehouseId, 1);
      expect(mockRepo.capturedDeleteLocationId, 5);
      expect(mockRepo.calls, ['deleteLocation']);
    });

    test('passes businessId to repository', () async {
      await useCases.deleteLocation(1, 5, businessId: 3);

      expect(mockRepo.capturedBusinessId, 3);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Delete failed');

      expect(() => useCases.deleteLocation(1, 5), throwsException);
    });
  });
}

File: front/central/src/services/modules/warehouses/app/use-cases.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { WarehouseUseCases } from './use-cases';
import { IWarehouseRepository } from '../domain/ports';
import {
    Warehouse,
    WarehouseDetail,
    WarehouseLocation,
    WarehousesListResponse,
    CreateWarehouseDTO,
    UpdateWarehouseDTO,
    CreateLocationDTO,
    UpdateLocationDTO,
} from '../domain/types';

// -----------------------------------------------------------------
// Helpers: datos de prueba reutilizables
// -----------------------------------------------------------------

const makeWarehouse = (overrides: Partial<Warehouse> = {}): Warehouse => ({
    id: 1,
    business_id: 1,
    name: 'Bodega Principal',
    code: 'BOD-001',
    address: 'Calle 100 #15-20',
    city: 'Bogota',
    state: 'Cundinamarca',
    country: 'CO',
    zip_code: '110111',
    phone: '3001234567',
    contact_name: 'Juan Perez',
    contact_email: 'juan@test.com',
    is_active: true,
    is_default: true,
    is_fulfillment: false,
    company: 'Test Company',
    first_name: 'Juan',
    last_name: 'Perez',
    email: 'juan@test.com',
    suburb: 'Centro',
    city_dane_code: '11001',
    postal_code: '110111',
    street: 'Calle 100',
    latitude: 4.6097,
    longitude: -74.0817,
    created_at: '2026-01-01T00:00:00Z',
    updated_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeLocation = (overrides: Partial<WarehouseLocation> = {}): WarehouseLocation => ({
    id: 1,
    warehouse_id: 1,
    name: 'Pasillo A',
    code: 'PA-001',
    type: 'shelf',
    is_active: true,
    is_fulfillment: false,
    capacity: 100,
    created_at: '2026-01-01T00:00:00Z',
    updated_at: '2026-01-01T00:00:00Z',
    ...overrides,
});

const makeWarehouseDetail = (overrides: Partial<WarehouseDetail> = {}): WarehouseDetail => ({
    ...makeWarehouse(),
    locations: [makeLocation()],
    ...overrides,
});

const warehousesListResponse: WarehousesListResponse = {
    data: [makeWarehouse()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
};

// -----------------------------------------------------------------
// Mock del repositorio
// -----------------------------------------------------------------

function createMockRepository(): IWarehouseRepository {
    return {
        getWarehouses: vi.fn(),
        getWarehouseById: vi.fn(),
        createWarehouse: vi.fn(),
        updateWarehouse: vi.fn(),
        deleteWarehouse: vi.fn(),
        getLocations: vi.fn(),
        createLocation: vi.fn(),
        updateLocation: vi.fn(),
        deleteLocation: vi.fn(),
    } as unknown as IWarehouseRepository;
}

// -----------------------------------------------------------------
// Suite principal
// -----------------------------------------------------------------

describe('WarehouseUseCases', () => {
    let repo: ReturnType<typeof createMockRepository>;
    let useCases: WarehouseUseCases;

    beforeEach(() => {
        repo = createMockRepository();
        useCases = new WarehouseUseCases(repo as unknown as IWarehouseRepository);
    });

    // ---------------------------------------------------------------
    // getWarehouses
    // ---------------------------------------------------------------
    describe('getWarehouses', () => {
        it('debería retornar la lista paginada de bodegas cuando el repositorio tiene éxito', async () => {
            vi.mocked(repo.getWarehouses).mockResolvedValue(warehousesListResponse);

            const result = await useCases.getWarehouses({ page: 1, page_size: 10 });

            expect(result).toEqual(warehousesListResponse);
            expect(repo.getWarehouses).toHaveBeenCalledOnce();
            expect(repo.getWarehouses).toHaveBeenCalledWith({ page: 1, page_size: 10 });
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan filtros', async () => {
            vi.mocked(repo.getWarehouses).mockResolvedValue(warehousesListResponse);

            await useCases.getWarehouses();

            expect(repo.getWarehouses).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Fallo de base de datos');
            vi.mocked(repo.getWarehouses).mockRejectedValue(expectedError);

            await expect(useCases.getWarehouses()).rejects.toThrow('Fallo de base de datos');
        });
    });

    // ---------------------------------------------------------------
    // getWarehouseById
    // ---------------------------------------------------------------
    describe('getWarehouseById', () => {
        it('debería retornar el detalle de una bodega por su ID', async () => {
            const detail = makeWarehouseDetail();
            vi.mocked(repo.getWarehouseById).mockResolvedValue(detail);

            const result = await useCases.getWarehouseById(1);

            expect(result).toEqual(detail);
            expect(repo.getWarehouseById).toHaveBeenCalledWith(1, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const detail = makeWarehouseDetail();
            vi.mocked(repo.getWarehouseById).mockResolvedValue(detail);

            await useCases.getWarehouseById(1, 5);

            expect(repo.getWarehouseById).toHaveBeenCalledWith(1, 5);
        });

        it('debería propagar el error cuando la bodega no existe', async () => {
            vi.mocked(repo.getWarehouseById).mockRejectedValue(new Error('Bodega no encontrada'));

            await expect(useCases.getWarehouseById(999)).rejects.toThrow('Bodega no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // createWarehouse
    // ---------------------------------------------------------------
    describe('createWarehouse', () => {
        const dto: CreateWarehouseDTO = {
            name: 'Bodega Nueva',
            code: 'BOD-002',
        };

        it('debería crear una bodega y retornar la respuesta del repositorio', async () => {
            const created = makeWarehouse({ id: 2, name: 'Bodega Nueva', code: 'BOD-002' });
            vi.mocked(repo.createWarehouse).mockResolvedValue(created);

            const result = await useCases.createWarehouse(dto);

            expect(result).toEqual(created);
            expect(repo.createWarehouse).toHaveBeenCalledOnce();
            expect(repo.createWarehouse).toHaveBeenCalledWith(dto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const created = makeWarehouse({ id: 2 });
            vi.mocked(repo.createWarehouse).mockResolvedValue(created);

            await useCases.createWarehouse(dto, 5);

            expect(repo.createWarehouse).toHaveBeenCalledWith(dto, 5);
        });

        it('debería propagar el error cuando la creación falla', async () => {
            vi.mocked(repo.createWarehouse).mockRejectedValue(new Error('Código duplicado'));

            await expect(useCases.createWarehouse(dto)).rejects.toThrow('Código duplicado');
        });
    });

    // ---------------------------------------------------------------
    // updateWarehouse
    // ---------------------------------------------------------------
    describe('updateWarehouse', () => {
        const updateDto: UpdateWarehouseDTO = { name: 'Bodega Actualizada', code: 'BOD-001' };

        it('debería actualizar una bodega y retornar la respuesta del repositorio', async () => {
            const updated = makeWarehouse({ name: 'Bodega Actualizada' });
            vi.mocked(repo.updateWarehouse).mockResolvedValue(updated);

            const result = await useCases.updateWarehouse(1, updateDto);

            expect(result).toEqual(updated);
            expect(repo.updateWarehouse).toHaveBeenCalledWith(1, updateDto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const updated = makeWarehouse();
            vi.mocked(repo.updateWarehouse).mockResolvedValue(updated);

            await useCases.updateWarehouse(1, updateDto, 5);

            expect(repo.updateWarehouse).toHaveBeenCalledWith(1, updateDto, 5);
        });

        it('debería propagar el error cuando la actualización falla', async () => {
            vi.mocked(repo.updateWarehouse).mockRejectedValue(new Error('Bodega no encontrada'));

            await expect(useCases.updateWarehouse(99, updateDto)).rejects.toThrow('Bodega no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // deleteWarehouse
    // ---------------------------------------------------------------
    describe('deleteWarehouse', () => {
        it('debería eliminar una bodega correctamente', async () => {
            vi.mocked(repo.deleteWarehouse).mockResolvedValue(undefined);

            await useCases.deleteWarehouse(1);

            expect(repo.deleteWarehouse).toHaveBeenCalledWith(1, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            vi.mocked(repo.deleteWarehouse).mockResolvedValue(undefined);

            await useCases.deleteWarehouse(1, 5);

            expect(repo.deleteWarehouse).toHaveBeenCalledWith(1, 5);
        });

        it('debería propagar la excepción cuando el repositorio lanza un error de red', async () => {
            vi.mocked(repo.deleteWarehouse).mockRejectedValue(new Error('Network error'));

            await expect(useCases.deleteWarehouse(1)).rejects.toThrow('Network error');
        });
    });

    // ---------------------------------------------------------------
    // getLocations
    // ---------------------------------------------------------------
    describe('getLocations', () => {
        it('debería retornar la lista de ubicaciones de una bodega', async () => {
            const locations = [makeLocation()];
            vi.mocked(repo.getLocations).mockResolvedValue(locations);

            const result = await useCases.getLocations(1);

            expect(result).toEqual(locations);
            expect(repo.getLocations).toHaveBeenCalledWith(1, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            vi.mocked(repo.getLocations).mockResolvedValue([]);

            await useCases.getLocations(1, 5);

            expect(repo.getLocations).toHaveBeenCalledWith(1, 5);
        });

        it('debería propagar el error cuando la consulta falla', async () => {
            vi.mocked(repo.getLocations).mockRejectedValue(new Error('Servicio no disponible'));

            await expect(useCases.getLocations(1)).rejects.toThrow('Servicio no disponible');
        });
    });

    // ---------------------------------------------------------------
    // createLocation
    // ---------------------------------------------------------------
    describe('createLocation', () => {
        const dto: CreateLocationDTO = { name: 'Pasillo B', code: 'PB-001' };

        it('debería crear una ubicación y retornar la respuesta del repositorio', async () => {
            const created = makeLocation({ id: 2, name: 'Pasillo B', code: 'PB-001' });
            vi.mocked(repo.createLocation).mockResolvedValue(created);

            const result = await useCases.createLocation(1, dto);

            expect(result).toEqual(created);
            expect(repo.createLocation).toHaveBeenCalledOnce();
            expect(repo.createLocation).toHaveBeenCalledWith(1, dto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const created = makeLocation({ id: 2 });
            vi.mocked(repo.createLocation).mockResolvedValue(created);

            await useCases.createLocation(1, dto, 5);

            expect(repo.createLocation).toHaveBeenCalledWith(1, dto, 5);
        });

        it('debería propagar el error cuando la creación falla', async () => {
            vi.mocked(repo.createLocation).mockRejectedValue(new Error('Código duplicado'));

            await expect(useCases.createLocation(1, dto)).rejects.toThrow('Código duplicado');
        });
    });

    // ---------------------------------------------------------------
    // updateLocation
    // ---------------------------------------------------------------
    describe('updateLocation', () => {
        const updateDto: UpdateLocationDTO = { name: 'Pasillo A Actualizado', code: 'PA-001' };

        it('debería actualizar una ubicación y retornar la respuesta del repositorio', async () => {
            const updated = makeLocation({ name: 'Pasillo A Actualizado' });
            vi.mocked(repo.updateLocation).mockResolvedValue(updated);

            const result = await useCases.updateLocation(1, 1, updateDto);

            expect(result).toEqual(updated);
            expect(repo.updateLocation).toHaveBeenCalledWith(1, 1, updateDto, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            const updated = makeLocation();
            vi.mocked(repo.updateLocation).mockResolvedValue(updated);

            await useCases.updateLocation(1, 1, updateDto, 5);

            expect(repo.updateLocation).toHaveBeenCalledWith(1, 1, updateDto, 5);
        });

        it('debería propagar el error cuando la actualización falla', async () => {
            vi.mocked(repo.updateLocation).mockRejectedValue(new Error('Ubicación no encontrada'));

            await expect(useCases.updateLocation(1, 99, updateDto)).rejects.toThrow('Ubicación no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // deleteLocation
    // ---------------------------------------------------------------
    describe('deleteLocation', () => {
        it('debería eliminar una ubicación correctamente', async () => {
            vi.mocked(repo.deleteLocation).mockResolvedValue(undefined);

            await useCases.deleteLocation(1, 1);

            expect(repo.deleteLocation).toHaveBeenCalledWith(1, 1, undefined);
        });

        it('debería pasar el businessId cuando se proporciona', async () => {
            vi.mocked(repo.deleteLocation).mockResolvedValue(undefined);

            await useCases.deleteLocation(1, 1, 5);

            expect(repo.deleteLocation).toHaveBeenCalledWith(1, 1, 5);
        });

        it('debería propagar la excepción cuando el repositorio lanza un error de red', async () => {
            vi.mocked(repo.deleteLocation).mockRejectedValue(new Error('Network error'));

            await expect(useCases.deleteLocation(1, 1)).rejects.toThrow('Network error');
        });
    });
});
